### DISCUSSION TITLE
Performance of materialized view with join

### DISCUSSION THREAD

**Discussion by atombender** [ClickHouse/ClickHouse#59187] [General]
I know that using joins in a materialized views are something of an antipattern. So I'd like to understand how to do this better.

My use case, simplified, is this. I have a table of requests and their latencies:

```sql
CREATE TABLE requests (
  timestamp DateTime('Etc/UTC'),
  customerId String,
  url String,
  params String,
  latency Float32,
  ... lots of additional columns ...
);
```

Then I have a view table that collects the average for each "group" of requests each day, which I call the baseline. The `key` here is derived from aspects of the requests that isn't relevant here:

```sql
CREATE TABLE agg_baselines (
  date Date,
  key String,
  stAvgLatency AggregateFunction(avg, Float32)
)
  ENGINE = AggregatingMergeTree
  PRIMARY KEY (date, key)
  ORDER BY (date, key);

CREATE MATERIALIZED VIEW agg_baselines_mv TO agg_baselines
SELECT
  date(timestamp) AS date,
  arrayStringConcat([customerId, url, params], '/') AS key, 
  avgState(latency) AS stAvgLatency
FROM requests
GROUP BY 1;
```

So far so good. While the `requests` table collects terabytes of data and several hundred million rows per day, the `agg_baselines` is nicely condensed into just a few megabytes and tends to be about 1m rows per day.

Now I want another view that summarizes per hour the number of requests that deviate from the baseline latency, grouped by `customerId` and `key`:

```sql
CREATE TABLE agg_regressions (
    timestamp DateTime('Etc/UTC'),
    customerId String,
    key String,
    stCountTotal AggregateFunction(count, UInt32),
    stCountRegression AggregateFunction(count, UInt32)
)
  ENGINE = AggregatingMergeTree
  PARTITION BY toYYYYMMDD(timestamp)
  PRIMARY KEY (toStartOfHour(timestamp), customerId, key)
  ORDER BY (toStartOfHour(timestamp), customerId, key);

CREATE MATERIALIZED VIEW agg_regressions_mv TO agg_regressions
SELECT
    toStartOfHour(timestamp) AS timestamp,
    customerId,
    arrayStringConcat([customerId, url, params], '/') AS key, 
    avgState(latency) AS stAvgLatency,
    countState(*) AS stCountTotal,
    -- Count above threshold derived from baseline
    countIfState(latency >= threshold) AS stCountRegression
FROM requests
INNER JOIN
(
    SELECT
        date,
        key,
        avgMerge(stAvgLatency) AS baseline,
        baseline * 1.5 AS threshold  -- 50% above baseline
    FROM agg_baselines
    GROUP BY 1, 2
) AS thresholds
  ON thresholds.key = key AND thresholds.date = (date(timestamp) - 1)
GROUP BY 1, 2, 3
```

The problem is that this MV, probably due to the join, is extremely slow, especially as the `agg_baselines` table builds up data over several days. My bulk inserts into `requests` are doing about 1m rows per minute, and with these two materialized views, bulk insert times go from ~15 seconds to ~6 minutes.

I have not partitioned the baseline table by date, as I hoped it would be unnecessary. Maybe it will help, but overall it seems like relying on materialized views for this type of aggregation is not the right solution.

I don't need real-time views; I would be happy with hourly or even daily updates. I looked at refreshable materialized views, but they appear to always rebuild the entire table from scratch on each run, and it's not practical or feasible to process a hundred terabytes of data every day. Unfortunately, a normal non-materialized view is too slow and memory-hungry to run over a period of more than a couple of days.

At the moment, the main alternative I'm considering is to not use MVs, but to manually populate these tables every day through a scheduled job or dataflow DAG. I don't know if there are any CH-native options.

Any ideas?

**Comment by den-crane** [comment]
Because Materialized View is an insert trigger the FROM table is the inserted buffer https://den-crane.github.io/Everything_you_should_know_about_materialized_views_commented.pdf

You can use this fact and limit/filter the right table

instead of
```
select ...
from A join B on id=id
....
```
you can do

```
select ...
from A join (select ... from B where B.id in (select A.id from A)) as B on id=id
....
```
`A` in this case not a real table in both cases of `from A`.

**Reply by atombender** [reply]
I'm familiar with that document, yes. What does your subquery trick accomplish that the join condition `ON thresholds.key = key AND thresholds.date = (date(timestamp) - 1)` does not? My materialized view operates on the buffer for the source table being inserted to, but it needs to join the entire baselines table (filtered by date and key).

**Comment by den-crane** [comment]
Another option is to replace JOIN with dictGet from external dictionary.

**Reply by atombender** [reply]
Dictionary might be an option, but it seems complicated? The baselines table stores potentially several months (several gigabytes) of data, and it needs to be incrementally built. Not sure how easy that is? I would be fine with building it daily, though.

**Reply by atombender** [reply]
I looked into the dictionary option. It may be feasible and even superior to using a MV for `agg_baselines`, because dictionaries are able to store a "window" of time thanks to expiration. Something like this, maybe:

```sql
 CREATE DICTIONARY baselines (
  key String,
  baseline Float32
)
  PRIMARY KEY (key)
  LAYOUT (HASHED())
  LIFETIME (MIN 0 MAX 3600)
  SOURCE(
    CLICKHOUSE(
      QUERY '
        SELECT
          arrayStringConcat([customerId, url, params], '/') AS key, 
          avg(latency) AS baseline
        FROM requests
        WHERE timestamp >= now() - INTERVAL 1 DAY
        GROUP BY 1
      '
    )
  )
```

This even allows a kind of sliding window over the underlying data used to compute baselines. The only problem here is that even this query can take about 20-30 seconds to run. I see that dictionaries are refreshed asynchronously, though it looks like the first `dictGet()` call is not.

I also see that `UPDATE_FIELD` and `UPDATE_LAG` could help refresh the table incrementally, but I'm not sure what happens with old rows, which we don't want to keep around.

**Comment by den-crane** [comment]
One more option is just store all this data in the same table with AggregatingMergeTree and compare two columns on the fly.

**Reply by atombender** [reply]
Not sure how this could be done with a single table without a join?

**Comment by ckevin** [comment]
I am late to this discussion, but a pattern that I have found effective is:
* Buffer the change data from the source table (`requests` in this case) through a NULL table.  That requires an extra table (the NULL) and an extra MV (from `requests` to NULL), but there is not much overhead to this and the benefit is that you can now reference the NULL table multiple times in a MV, allowing you to use the new block in different ways within the same query.  If instead you triggered off `requests`, it would only represent the new block the first time you use that symbol -- each subsequent usage would reference the whole `requests` table instead.
* Use the NULL table as the trigger for your MV.
* Use that NULL data set in multiple places, both as the left side of your join, but also to limit what data you join to from `agg_baselines`.  I am guessing this is your expensive part, since the `SELECT` from `agg_baselines` is not limited at all:  
    ```
    (
    SELECT
        date,
        key,
        avgMerge(stAvgLatency) AS baseline,
        baseline * 1.5 AS threshold  -- 50% above baseline
    FROM agg_baselines
    GROUP BY 1, 2
    ) AS thresholds
    ```
* Use subquery joins, with `IN (...)` clauses based on a prefix of the `ORDER BY` clause of the joined table (or projection), to make sure you get the best incremental performance.

Putting this together, I would expect something in the spirit of the following MV to perform better for you.  Note that the subquery filter uses the same criteria as the outer join to make sure your `thresholds` data set has all the necessary rows (`date - 1` for example):
```
CREATE MATERIALIZED VIEW agg_regressions_mv TO agg_regressions
SELECT
...
FROM null_requests
INNER JOIN
(
    SELECT
        date,
        key,
        avgMerge(stAvgLatency) AS baseline,
        baseline * 1.5 AS threshold  -- 50% above baseline
    FROM agg_baselines
    WHERE (date, key) IN (
        SELECT date(timestamp) - 1, arrayStringConcat([customerId, url, params], '/')
        FROM null_requests
    )
    GROUP BY 1, 2
) AS thresholds
    ON thresholds.key = key AND thresholds.date = (date(timestamp) - 1)
GROUP BY 1, 2, 3
```
If your inserts to `requests` tend to only involve recent data, then the amount of time you pull from `agg_baselines` at each MV invocation should not grow over time with this strategy, due to the new `WHERE` filter.  You may not even need the `key` part on the `WHERE` clause, I'd test whether it is necessary.
